Google.Api.Generator.Utils/SourceFileContext.cs

```diff
@@ -103,55 +103,23 @@ private sealed class Unaliased : SourceFileContext
-            // Name collisions between types imported from two different namespaces are not handled;
-            // however, this isn't necessary as we completely control which namespaces are imported,
-            // and we know that there are no collisions.
-
-            public Unaliased(IClock clock) : base(clock) { }
-
-            // Imported namespaces.
-            private readonly HashSet<string> _imports = new HashSet<string>();
-            // All type-names that have been imported through namespace imports. Used to detect name collisions.
-            private readonly HashSet<string> _importedClassNames = new HashSet<string>();
+            public Unaliased(IClock clock, IReadOnlyDictionary<string, string> wellKnownNamespaceAliases) : base(clock) =>
+                _wellKnownNamespaceAliases = wellKnownNamespaceAliases;
 
-            private void AddImport(string ns)
-            {
-                // Add import.
-                if (_imports.Add(ns))
-                {
-                    // Track all available type names that can be used unalised.
-                    // This mechanism isn't fool-proof, but will work fine in most cases because this generator
-                    // depends on all assemblies that a generated library will depend on.
-                    var typeNamesInNs = AppDomain.CurrentDomain.GetAssemblies()
-                        .Where(a => !a.IsDynamic)
-                        .SelectMany(a => a.GetExportedTypes())
-                        .Where(t => t.Namespace == ns)
-                        .Select(t => t.Name);
-                    foreach (var typeNameInNs in typeNamesInNs)
-                    {
-                        _importedClassNames.Add(typeNameInNs);
-                    }
-                }
-            }
+            private readonly IReadOnlyDictionary<string, string> _wellKnownNamespaceAliases;
+            // Seen namespaces. The associated bool value indicates whether we may skip import or not.
+            private readonly Dictionary<string, bool> _seenNamespaces = new Dictionary<string, bool>();
+            // Seen types and in which distinct namespaces. This will be used to avoid name collisions.
+            private readonly Dictionary<string, HashSet<string>> _seenTypeNames = new Dictionary<string, HashSet<string>>();
 
             public override TypeSyntax Type(Typ typ) => base.Type(typ) ?? Type0(typ);
 
-            private const string AliasableType = "rewriteable-type";
+            private const string FromNamespace = "from-namespace";
 
             public TypeSyntax Type0(Typ typ)
             {
-                bool aliasable;
-                if (!$"{Namespace}.".StartsWith($"{typ.Namespace}."))
-                {
-                    AddImport(typ.Namespace);
-                    aliasable = false;
-                }
-                else
-                {
-                    aliasable = true;
-                }
+                Seen(typ);
```

Comment:
Possibly "mark as seen" or "update seen state" or similar? It would be nice to have an actual verb in there :)